pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"math"
+	"net"
+	"sync"
+)
+
+const (
+	ipRangeSize = 29
+	byteMax     = 255
+	ipV4Bits    = 32
+)
+
+// IPAllocator struct consists of shared resources that are used to keep track of the /29 IPRanges currently reserved by service instances
+type IPAllocator struct {
+	// pendingIPRanges set maintains the set of  IP ranges that have been reserved by the service instances but pending reservation in the cloud instances
+	// The key is a IP range currently reserved by a service instance e.g(192.168.92.0/29). Value is a bool to implement map as a set
+	pendingIPRanges map[string]bool
+
+	// pendingIPRangesMutex is used to synchronize access to the pendingIPRanges set to prevent data races
+	pendingIPRangesMutex sync.Mutex
+}
+
+// NewIPAllocator is the constructor to initialize the IPAllocator object
+// Argument pendingIPRanges map[string]bool is a set of  IP ranges currently reserved by service instances but pending reservation in the cloud instances
+func NewIPAllocator(pendingIPRanges map[string]bool) *IPAllocator {
+	return &IPAllocator{
+		pendingIPRanges: pendingIPRanges,
```

Comment:
Instead of setting it to the same map, I would make a copy of the argument.  The reason is the caller has the ability to mutate the map outside of this library, which could interfere with the functions inside.